datalad/interface/run.py

```diff
@@ -0,0 +1,268 @@
+
+lgr = logging.getLogger('datalad.interface.run')
+
+
+@build_doc
+class Run(Interface):
+    """Run and explain
+
+    If the executed command did not alter the dataset in any way, no record of
+    the command execution is made.
+
+    If the given command errors, a `CommandError` exception with the same exit
+    code will be raised.
+
+    The executed command, the process working directory (PWD; if it is inside
+    the given dataset), and the command's exit code will be recorded in the
+    commit message.
+    """
+    _params_ = dict(
+        cmd=Parameter(
+            args=("cmd",),
+            nargs=REMAINDER,
+            metavar='SHELL COMMAND',
+            doc="command for execution"),
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc="""specify the dataset to query.  If
+            no dataset is given, an attempt is made to identify the dataset
+            based on the input and/or the current working directory""",
+            constraints=EnsureDataset() | EnsureNone()),
+        message=save_message_opt,
+        rerun=Parameter(
+            args=('--rerun',),
+            action='store_true',
+            doc="""re-run the command recorded in the last saved change (if any).
+            This will ignore any command given as an argument, and execute the
+            recorded command call in the recorded working directory. The recorded
+            changeset will be replaced by the outcome of the command re-run."""),
+        # TODO
+        # --list-commands
+        #   go through the history and report any recorded command. this info
+        #   could be used to unlock the associated output files for a rerun
+        # --rerun #
+        #   rerun command # from the list of recorded commands using the info/cmd
+        #   on record
+    )
+
+    @staticmethod
+    @datasetmethod(name='run')
+    @eval_results
+    def __call__(
+            # it is optional, because `rerun` can get a recorded one
+            cmd=None,
+            dataset=None,
+            message=None,
+            rerun=False):
+        if rerun and cmd:
+            lgr.warning('Ignoring provided command in --rerun mode')
+            cmd = None
+        if not dataset:
+            # act on the whole dataset if nothing else was specified
+            dataset = get_dataset_root(curdir)
+        ds = require_dataset(
+            dataset, check_installed=True,
+            purpose='tracking outcomes of a command')
+        # not needed ATM
+        #refds_path = ds.path
+        lgr.debug('tracking command output underneath %s', ds)
+        try:
+            # base assumption is that the animal smells superb
+            ok_clean_git(ds.path)
+        except AssertionError:
+            yield get_status_dict(
+                'run',
+                ds=ds,
+                status='impossible',
+                message='unsaved modifications present, cannot detect changes by command')
+            return
```

Comment:
ATM would not be able to run whenever I have some files unlocked (e.g. to be modified):
```shell
(git-annex)hopa:/tmp/ds[master]git
$> datalad unlock 123                                 
unlock(ok): 123 (file)
$> datalad run "sed  -e 's,9,1,g' 123 | sponge > 123"
run(impossible): /tmp/ds (dataset) [unsaved modifications present, cannot detect changes by command]
```
```shell
$> git status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	typechange: 123
```
making it not usable for any scenario where command should modify/(often as well overwrite if write is done without removing previous one first) an already existing file